File: Programmers/Lv3/단어 변환.js
function solution(begin, target, words) {
  var answer = Infinity;
  if (!words.includes(target)) {
    return 0;
  }

  const visited = new Set();

  const dfs = (s, cnt) => {
    if (s === target) {
      answer = Math.min(cnt, answer);
    }

    for (let word of words) {
      if (!visited.has(word) && cvt(s, word)) {
        visited.add(word);
        dfs(word, cnt + 1);
        visited.delete(word);
      }
    }
  };

  const cvt = (s1, s2) => {
    let diff = 0;
    for (let i = 0; i < s1.length; i++) {
      if (s1[i] !== s2[i]) {
        diff++;
      }
    }

    if (diff === 1) {
      return true;
    } else if (diff > 1) {
      return false;
    }
  };

  dfs(begin, 0);
  return answer;
}

File: Programmers/Lv3/단어 변환.cpp
#include <bits/stdc++.h>

using namespace std;

int solution(string begin, string target, vector<string> words) {
    int n = begin.length();
    set<string> visited;
    queue<pair<string, int>> q;
    
    visited.insert(begin);
    q.push({begin, 0});
    
    while(!q.empty()){
        auto tmp = q.front();
        string node = tmp.first;
        int cnt = tmp.second;
        
        q.pop();
        
        if(node == target){
            return cnt;
        }
        
        for(string& word : words){
            if(visited.find(word) == visited.end()){
                int diff = 0;
            
                for(int i = 0; i < n; i++){
                    if(node[i] != word[i]){
                        diff++;
                    }
                }
                if(diff == 1){
                    q.push({word, cnt + 1});
                    visited.insert(word);
                }   
            }
        }

    }
    return 0;
}